File: dashboard/src/lib/skillGate.ts
/* ---------------------------------------------------------------------------
   Operator skill kill-switch — the console side of the /v1/admin/skills endpoints.

   Disabling a skill is a DENY-only control: while disabled, an alias is denied
   SKILL_DISABLED for every caller in the tenant, regardless of capability. It
   NEVER edits the alias→target mapping — the obfuscation layer stays immutable;
   the operator can only toggle availability.

   Runs under a CAP_DIRECTORY_ADMIN token minted for the catalog's own tenant
   (disable is tenant-scoped).
--------------------------------------------------------------------------- */

import {
  listDisabledSkills,
  listRegisteredSkills,
  setSkillDisabled as apiSetSkillDisabled,
  registerSkill as apiRegisterSkill,
  deregisterSkill as apiDeregisterSkill,
  mintDevToken,
  type RegisterSkillBody,
  type RegisteredSkill,
} from './api';
import { CAP_DIRECTORY_ADMIN } from './protocol';

async function adminToken(apiBase: string, tenantId: string, signal?: AbortSignal): Promise<string | null> {
  try {
    return await mintDevToken(
      { tenant_id: tenantId, agent_id: 'agent-directory-admin', capabilities: [CAP_DIRECTORY_ADMIN] },
      { base: apiBase, signal },
    );
  } catch {
    return null;
  }
}

/** Load the set of disabled alias names for the tenant. */
export async function loadDisabledSkills(apiBase: string, tenantId: string, signal?: AbortSignal): Promise<string[]> {
  const token = await adminToken(apiBase, tenantId, signal);
  if (!token) return [];
  return (await listDisabledSkills(token, { base: apiBase, signal })) ?? [];
}

/** Load the operator-registered (deregisterable) skills — alias + creation timestamp. */
export async function loadRegisteredSkills(apiBase: string, tenantId: string, signal?: AbortSignal): Promise<RegisteredSkill[]> {
  const token = await adminToken(apiBase, tenantId, signal);
  if (!token) return [];
  return (await listRegisteredSkills(token, { base: apiBase, signal })) ?? [];
}

/** Disable or enable one alias for the tenant. Returns true on a durable change. */
export async function setSkillDisabled(
  apiBase: string,
  tenantId: string,
  alias: string,
  disabled: boolean,
  signal?: AbortSignal,
): Promise<boolean> {
  const token = await adminToken(apiBase, tenantId, signal);
  if (!token) return false;
  return apiSetSkillDisabled(token, alias, disabled, { base: apiBase, signal });
}

/**
 * Register a NEW operator skill (a new alias→target) for the tenant. Additive only —
 * the gateway refuses to shadow a config alias. Returns true on a durable 200.
 */
export async function registerSkill(
  apiBase: string,
  tenantId: string,
  body: RegisterSkillBody,
  signal?: AbortSignal,
): Promise<boolean> {
  const token = await adminToken(apiBase, tenantId, signal);
  if (!token) return false;
  return apiRegisterSkill(token, body, { base: apiBase, signal });
}

/**
 * Deregister an OPERATOR-registered skill. `removed` is true only when an overlay row
 * was actually dropped (config aliases are a no-op success). Returns { ok, removed }.
 */
export async function deregisterSkill(
  apiBase: string,
  tenantId: string,
  alias: string,
  signal?: AbortSignal,
): Promise<{ ok: boolean; removed: boolean }> {
  const token = await adminToken(apiBase, tenantId, signal);
  if (!token) return { ok: false, removed: false };
  return apiDeregisterSkill(token, alias, { base: apiBase, signal });
}

File: services/skill_gate.py
"""
MCPIP V2 — Service: SkillGateStore (operator skill kill-switch).

    ◐ "Disable a skill for everyone; never touch where it points."

An admin holding ``CAP_DIRECTORY_ADMIN`` can DISABLE an alias for their tenant.
While disabled, every invocation of that alias is denied ``SKILL_DISABLED`` —
regardless of the caller's capabilities — until an admin re-enables it. This is
the skill-side twin of the principal kill-switch (``RevocationStore``): a fast,
reversible operator control to take a tool offline (a compromised integration, a
deprecated skill, an incident) without a redeploy.

It NEVER edits the alias→target mapping. The obfuscation layer (which real system
an alias resolves to) stays immutable config; this store only holds a per-tenant
set of *disabled alias names* and the pipeline turns membership into a DENY. So a
console operator can toggle availability without ever being able to repoint a
skill at a different (or malicious) target.

Fail-closed discipline (mirrors ``RevocationStore``):
  * ``is_disabled`` runs on the hot path; a Redis transport failure raises
    ``LockError`` (→ LOCK_ERROR deny) so an unreadable disable-set never lets a
    possibly-disabled skill through.
  * ``disable`` / ``enable`` are admin mutations; a transport failure raises
    ``LockError`` so the operator learns the toggle did not durably land.
"""

from __future__ import annotations

from typing import Any

import redis.asyncio as redis
from redis.exceptions import RedisError

from auth import LockError

_KEY_PREFIX = "mcpip:skill_disabled:"


class SkillGateStore:
    """Redis-backed per-tenant set of disabled alias names — fail-closed reads."""

    def __init__(self, redis_client: "redis.Redis") -> None:
        self._redis = redis_client

    @staticmethod
    def _key(tenant_id: str) -> str:
        # Tenant-scoped: an admin disables/enables ONLY within its own tenant, and
        # ``is_disabled`` is queried with the token's own tenant, so a disable can
        # never leak across the tenant boundary.
        return f"{_KEY_PREFIX}{tenant_id}"

    async def disable(self, tenant_id: str, alias: str) -> bool:
        """Disable ``alias`` for ``tenant_id``. Returns True iff newly added."""
        try:
            added: Any = await self._redis.sadd(self._key(tenant_id), alias)
        except RedisError as exc:
            raise LockError("skill gate transport failure during disable") from exc
        return int(added) > 0

    async def enable(self, tenant_id: str, alias: str) -> bool:
        """Re-enable ``alias`` for ``tenant_id``. Returns True iff it was disabled."""
        try:
            removed: Any = await self._redis.srem(self._key(tenant_id), alias)
        except RedisError as exc:
            raise LockError("skill gate transport failure during enable") from exc
        return int(removed) > 0

    async def is_disabled(self, tenant_id: str, alias: str) -> bool:
        """
        True iff ``alias`` is disabled for ``tenant_id``. Runs on the hot path.

        Fail-closed: a transport failure raises ``LockError`` (mapped to a
        LOCK_ERROR deny by the funnel) — an unreadable disable-set never lets a
        possibly-disabled skill through.
        """
        try:
            member: Any = await self._redis.sismember(self._key(tenant_id), alias)
        except RedisError as exc:
            raise LockError("skill gate transport failure") from exc
        return bool(member)

    async def list_disabled(self, tenant_id: str) -> list[str]:
        """
        Return the disabled alias names for ``tenant_id`` (operator view). Fail-soft:
        a transport error yields an empty list rather than raising — this backs a
        read-only listing, never an authorization decision.
        """
        try:
            members: Any = await self._redis.smembers(self._key(tenant_id))
        except RedisError:
            return []
        return sorted(m.decode() if isinstance(m, bytes) else str(m) for m in members)


__all__ = ["SkillGateStore"]
